FSW/src/util/memUtil.cpp

```diff
@@ -0,0 +1,250 @@
+static float dataBuffer[BUFFERSIZE]; // create array to hold data buffer elements
+
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - dataProcessing - - - - - - *
+ * Usage:
+ *  Reads incoming data from ADC, measures against baseline,
+ *  sends to memory.
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  data from the ADC (voltage measurement)
+ */
+float dataProcessing()
+{
+    float voltage;
+
+    // begin an SPI connection
+    // TODO: we should probably begin the SPI connection in main to avoid duplicates
+    SPI.begin();  
+    pinMode(ADC_CHIP_SELECT, OUTPUT); // set ADC chip select pin to output
+
+    // access ADC Pin (SPI)
+    uint16_t photodiode16; 
+    SPI.beginTransaction(SPISettings(SPI_MAX_SPEED, MSBFIRST, SPI_MODE3)); //SPISettings(maxSpeed,dataOrder,dataMode)
+    digitalWrite(ADC_CHIP_SELECT, LOW);   // set Slave Select pin to low to select chip
+    photodiode16 = SPI.transfer16(0x0000);// transfer data, send 0 to slave, recieve data from ADC
+    digitalWrite(ADC_CHIP_SELECT, HIGH);  // set Slave Select pin to high to de-select chip
+    SPI.endTransaction();
+
+    // Convert ADC output to voltage by scaling over voltage range
+
+    //TODO: can we scale this by the actual digital voltage on the board?
+    voltage = photodiode16 / ADC_BINS * (ADC_MAX_VOLTAGE - ADC_MIN_VOLTAGE);
+
+    return voltage;
+}
+
+/* - - - - - - scienceMemoryHandling - - - - - - *
+ * Usage:
+ *  Takes data from processing and stores it in a buffer,
+ *  continuously overwriting old entries.
+ *  Copies buffer to long term memory when signaled by timing module.
+ * 
+ * Inputs:
+ *  none
+ *  
+ * Outputs:
+ *  None
+ */
+void scienceMemoryHandling()
+{    
+    if (dataProcessEvent.checkInvoked()){ // checks event status from timing module
+        float currentPhotodiodeVoltage = dataProcessing();
+
+        Serial.print("Current Photodiode Voltage: ");
+        Serial.print(currentPhotodiodeVoltage);
+        Serial.print("\n");
+
+        updateBuffer(currentPhotodiodeVoltage, bufIdx);
+    }
+
+    if (saveEvent.checkInvoked()){
+        saveBuffer(bufIdx);
+    }
+}
+
+/* - - - - - - Helper Functions - - - - - - */
+
+/* - - - - - - updateBuffer - - - - - - *
+ * Belongs to Science Memory Handling Module
+ *  
+ * Usage:
+ *  adds a new sample to the data buffer at specified index
+ *  an existing element at index will be overwritten
+ * 
+ * Inputs:
+ *  dataBuffer - pointer to first element of data buffer
+ *  sample - data to be stored (type must match dataBuffer type)
+ *  index - index to store new data at (pass-by-reference)
+ *  
+ * Outputs:
+ *  none
+ */
+void updateBuffer(float sample, int &index){
+    
+    // check that index is valid
+    if (0 <= index && index < BUFFERSIZE){
+        dataBuffer[index] = sample;
+        index++;
+    
+    } else {
+        Serial.print("WARNING: invalid dataBuffer index ");
+        Serial.print("(Science Memory Handling Module)\n");
```

Comment:
This is great! Indicating the module that prints a warning makes debugging and testing much easier.

Reply:
Yea I agree, I think you two should use this pattern elsewhere. Good example to follow with temperature checks and stuff. I think FSW should have automated triggers to enter safe modes, extending the functionality of these check blocks.